### PR TITLE
c8y-mapper consume the "te" new alarm mqtt scheme

### PR DESCRIPTION
**Pull request by rina23q**
## Proposed changes
Implement #2243.

Behaviour changes:
- Defined the default values for mandatory fields for Cumulocity if it's missing or wrongly given.
a. type: `ThinEdgeAlarm`
b. text: The alarm type (the same as previous behaviour)
c. severity: `minor`
- The definition of uniqueness of alarm. Now it's per alarm type, although it was the combination of alarm type and severity previously. So, alarm A `"type": "foo", "severity": "major"` and alarm B `"type": "foo", "severity": "critical"` are no longer considered as different alarms.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
#2243 

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: crates/core/tedge_api/src/alarm.rs [hunk 1]
```diff
@@ -1,58 +1,84 @@
...
+    pub alarm_type: Option<String>,
 
-#[derive(Debug, Serialize, Deserialize, Eq, PartialEq)]
-pub enum AlarmSeverity {
-    Critical,
-    Major,
-    Minor,
-    Warning,
+    /// Alarm source retrieved from the MQTT topic.
+    pub source: EntityMetadata,
```
Comment (didier-wenzek) [inline crates/core/tedge_api/src/alarm.rs:17]: I think it would be more appropriate to use an [`EntityTopicId`](https://github.com/thin-edge/thin-edge.io/blob/main/crates/core/tedge_api/src/mqtt_topics.rs#L213) instead of an [`EntityMetadata`](https://github.com/thin-edge/thin-edge.io/blob/main/crates/core/tedge_api/src/entity_store.rs#L288).

- An `EntityTopicId` is just an identifier such as `device/xyz//`, while an `EntityMetadata` contains all the data attached to that entity.
- Using an `EntityMetadata` to identify the source of an alarm makes things a bit more involved (one needs an [`EntityStore`](https://github.com/thin-edge/thin-edge.io/blob/main/crates/core/tedge_api/src/entity_store.rs#L58) to decode the `EntityTopicId` extracted from the alarm topic.
- One can not represent an alarm sent by a device that is not registered yet (as it is then unknown from the `EntityStore`).

File: crates/core/tedge_api/src/alarm.rs [hunk 2]
```diff
@@ -1,58 +1,84 @@
...
+            data: None,
+        }
+    }
 
-    #[error("Unsupported alarm severity in topic: {0}")]
-    UnsupportedAlarmSeverity(String),
+    /// Override "type" when it is provided in payload.
+    pub fn with_alarm_data(self, mut data: ThinEdgeAlarmData) -> Self {
+        match data.extras.remove("type") {
+            Some(maybe_type_from_payload) => match maybe_type_from_payload.as_str() {
```
Comment (didier-wenzek) [inline crates/core/tedge_api/src/alarm.rs:52]: Why do we need this second match ` maybe_type_from_payload.as_str()` ?

File: crates/core/c8y_api/src/json_c8y.rs [hunk 3]
```diff
@@ -236,95 +238,167 @@ impl SourceInfo {
...
-        Ok(Self {
-            source,
-            severity,
-            alarm_type,
-            time,
-            text,
-            fragments,
-        })
+    // TODO: Ask which one is better, using default or returning an error.
+    fn convert_severity(alarm_data: &ThinEdgeAlarmData) -> AlarmSeverity {
```
Comment (didier-wenzek) [inline crates/core/c8y_api/src/json_c8y.rs:325]: In that case, using a default value is appropriate and convenient for the users.

File: crates/extensions/c8y_mapper_ext/src/alarm_converter.rs [hunk 4]
```diff
@@ -33,62 +34,67 @@ impl AlarmConverter {
...
-                    output_messages.push(Message::new(&c8y_alarm_topic, cumulocity_alarm_json));
+                match c8y_alarm.data {
+                    Some(ref alarm_data) if !alarm_data.fragments.is_empty() => {
+                        // JSON over MQTT
+                        let cumulocity_alarm_json = serde_json::to_string(&c8y_alarm)?;
+                        let c8y_alarm_topic = Topic::new_unchecked(C8Y_JSON_MQTT_ALARMS_TOPIC);
+                        output_messages.push(Message::new(&c8y_alarm_topic, cumulocity_alarm_json));
+                    }
+                    _ => {
+                        // SmartREST
```
Comment (didier-wenzek) [inline crates/extensions/c8y_mapper_ext/src/alarm_converter.rs:87]: Why not sending all the alarms using C8Y JSON?

File: crates/extensions/c8y_mapper_ext/src/alarm_converter.rs [hunk 5]
```diff
@@ -33,62 +34,67 @@ impl AlarmConverter {
...
+                    let tedge_alarm_data =
+                        ThinEdgeAlarmData::try_from(mqtt_payload).map_err(|e| {
+                            ThinEdgeAlarmDeserializerError::FailedToParseJsonPayload {
+                                topic: mqtt_topic,
+                                error: e.to_string(),
+                                payload: mqtt_payload.chars().take(50).collect(),
+                            }
+                        })?;
+                    ThinEdgeAlarm::new(alarm_type, source).with_alarm_data(tedge_alarm_data)
+                };
```
Comment (didier-wenzek) [inline crates/extensions/c8y_mapper_ext/src/alarm_converter.rs:74]: This logic should be in `tedge_api::alarms` as this is unrelated to Cumulocity.

File: crates/core/tedge_api/src/alarm.rs [hunk 6]
```diff
@@ -1,58 +1,84 @@
...
 pub struct ThinEdgeAlarm {
-    pub name: String,
-    pub severity: AlarmSeverity,
-    #[serde(flatten)]
-    pub data: Option<ThinEdgeAlarmData>,
-    pub source: Option<String>,
-}
+    /// Alarm type retrieved from the MQTT topic or payload.
+    /// The "type" given from the payload has priority to the one from the topic.
+    pub alarm_type: Option<String>,
```
Comment (didier-wenzek) [inline crates/core/tedge_api/src/alarm.rs:14]: Why not a `String` as the empty string is a valid alarm type (for thin-edge).

File: crates/core/tedge_api/src/alarm.rs [hunk 7]
```diff
@@ -1,58 +1,84 @@
...
-
 use serde_json::Value;
 use std::collections::HashMap;
-/// In-memory representation of ThinEdge JSON alarm.
-#[derive(Debug, Serialize, Deserialize, Eq, PartialEq)]
+use std::convert::TryFrom;
+
+/// Internal representation of ThinEdge alarm model.
+#[derive(Debug, Eq, PartialEq)]
 pub struct ThinEdgeAlarm {
```
Comment (didier-wenzek) [inline crates/core/tedge_api/src/alarm.rs:11]: I really like this decomposition of an alarm in two parts: one extracted from the topic, the other from the payload.

- My initial idea was to have a single structure, but I agree that this leads to issues when parsing an MQTT message.
- Using an optional payload makes clear and less error prone the case of the empty messages used to clear alarms.

That said it would be good to find a generic approach here. The same pattern `(source, type, payload)` is used for alarms, events, measurements, commands ...

**Review by didier-wenzek — COMMENTED** [review]
I have some questions

File: crates/core/tedge_api/src/alarm.rs (hunk 2 above)
Comment (rina23q) [reply crates/core/tedge_api/src/alarm.rs:52]: The hashmap `data.extras` may have `type: true` (boolean), not the desirable type. In this case, `as_str()` returns `Err`. If it's `Err`, I want to use the default type `ThinEdgeAlarm`.

File: crates/extensions/c8y_mapper_ext/src/alarm_converter.rs (hunk 4 above)
Comment (rina23q) [reply crates/extensions/c8y_mapper_ext/src/alarm_converter.rs:87]: Agree! Still need to check how to clear an alarm by JSON over MQTT though. But I would address this later with https://github.com/thin-edge/thin-edge.io/issues/2242.

File: crates/extensions/c8y_mapper_ext/src/alarm_converter.rs (hunk 5 above)
Comment (rina23q) [reply crates/extensions/c8y_mapper_ext/src/alarm_converter.rs:74]: You are right. The logic should move to the `ThinEdgeAlarm` side.

File: crates/core/tedge_api/src/alarm.rs (hunk 6 above)
Comment (rina23q) [reply crates/core/tedge_api/src/alarm.rs:14]: Actually it should be `String`. I was overthinking about the case like `te/device/main///a/` (empty type), but this must be `"".to_string()`, not `None`. In short, I'll fix it.

File: crates/extensions/c8y_mapper_ext/src/alarm_converter.rs [hunk 8]
```diff
@@ -33,62 +34,67 @@ impl AlarmConverter {
...
-                            payload: mqtt_payload.chars().take(50).collect(),
-                        }
-                    })?;
-                let c8y_alarm = C8yCreateAlarm::try_from(&tedge_alarm)?;
+                let tedge_alarm = if mqtt_payload.is_empty() {
+                    // Clearing an alarm
+                    ThinEdgeAlarm::new(alarm_type, source)
+                } else {
+                    // Creating an alarm
+                    let tedge_alarm_data =
```
Comment (albinsuresh) [inline crates/extensions/c8y_mapper_ext/src/alarm_converter.rs:65]: I'd just restore that old `ThinEdgeAlarm::try_from` function or something like `impl TryFrom<&Message> for ThinEdgeAlarm`

File: crates/extensions/c8y_mapper_ext/src/alarm_converter.rs (hunk 4 above)
Comment (albinsuresh) [reply crates/extensions/c8y_mapper_ext/src/alarm_converter.rs:87]: Just legacy, I guess. But doing that would be much better, in terms of code complexity, as then we just need to create one `C8yCreateAlarm` structure and use that both for MQTT as well as HTTP.

File: crates/extensions/c8y_mapper_ext/src/converter.rs [hunk 9]
```diff
@@ -318,33 +318,19 @@ impl CumulocityConverter {
...
+            let mqtt_messages = self
+                .alarm_converter
+                .try_convert_alarm(entity, input, alarm_type)?;
             Ok(mqtt_messages)
-        } else if topic.name.starts_with(INTERNAL_ALARMS_TOPIC) {
-            self.alarm_converter.process_internal_alarm(message);
-            Ok(vec![])
         } else {
-            Err(ConversionError::UnsupportedTopic(topic.name.clone()))
+            Ok(vec![])
```
Comment (albinsuresh) [inline crates/extensions/c8y_mapper_ext/src/converter.rs:333]: I'd return an error about unknown entity so that this error is properly reported on `tedge/errors`.

File: crates/core/tedge_api/src/alarm.rs [hunk 10]
```diff
@@ -1,58 +1,84 @@
...
+    /// All other alarm data retrieved from the MQTT payload.
+    /// None means the message is meant to clear an alarm.
+    pub data: Option<ThinEdgeAlarmData>,
 }
 
-/// In-memory representation of ThinEdge JSON alarm payload
+/// Internal representation of the JSON MQTT payload.
 #[derive(Debug, Serialize, Deserialize, Eq, PartialEq)]
 pub struct ThinEdgeAlarmData {
+    pub severity: Option<String>,
```
[the remark this answers is not in the document]
Comment (rina23q) [reply crates/core/tedge_api/src/alarm.rs:27]: It's thin-edge level. I thought the new scheme would allow full freedom for any fields. Rather, why do we need to have limited `severity` as Cumulocity?

File: tests/RobotFramework/tests/cumulocity/telemetry/thin-edge_device_telemetry.robot [hunk 11]
```diff
@@ -69,12 +69,50 @@ Thin-edge devices support sending custom events without type in topic
...
 Thin-edge devices support sending custom alarms #1699
     [Tags]    \#1699
-    Execute Command    tedge mqtt pub tedge/alarms/critical/myCustomAlarmType '{ "text": "Some test alarm", "someOtherCustomFragment": {"nested":{"value": "extra info"}} }'
+    Execute Command    tedge mqtt pub te/device/main///a/myCustomAlarmType '{ "severity": "critical", "text": "Some test alarm", "someOtherCustomFragment": {"nested":{"value": "extra info"}} }'
     ${alarms}=    Device Should Have Alarm/s    expected_text=Some test alarm    severity=CRITICAL    minimum=1    maximum=1    type=myCustomAlarmType
     Should Be Equal    ${alarms[0]["someOtherCustomFragment"]["nested"]["value"]}    extra info
     Log    ${alarms}
 
 
+Thin-edge devices support sending custom alarms overriding the type
```
[the remark this answers is not in the document]
Comment (rina23q) [reply tests/RobotFramework/tests/cumulocity/telemetry/thin-edge_device_telemetry.robot:78]: Honestly I don't know about the use-cases. But this is the same way as we support for measurements and events.

File: crates/core/c8y_api/src/json_c8y.rs [hunk 12]
```diff
@@ -236,95 +238,167 @@ impl SourceInfo {
...
+    }
+
+    /// Remove reserved keywords from extras.
+    /// "type", "time", "text", "severity" are ensured that they are not
+    /// in the hashmap of ThinEdgeAlarm because they are already members of the struct itself.
+    fn convert_extras(alarm_data: &ThinEdgeAlarmData) -> HashMap<String, Value> {
+        let mut map = alarm_data.extras.clone();
+        map.remove("externalSource");
+        map
     }
```
[the remark this answers is not in the document]
Comment (rina23q) [reply crates/core/c8y_api/src/json_c8y.rs:350]: These functions are private and called by `From`. Technically, we can break these functions but I didn't want to have very long implementation inside `From`.

File: crates/core/c8y_api/src/json_c8y.rs [hunk 13]
```diff
@@ -236,95 +238,167 @@ impl SourceInfo {
...
     #[serde(skip_serializing_if = "Option::is_none")]
     #[serde(rename = "externalSource")]
     pub source: Option<SourceInfo>,
 
-    pub severity: String,
+    /// If None, the message is meant to clear the alarm,
+    /// Some means to create a new alarm.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[serde(flatten)]
+    pub data: Option<C8yCreateAlarmData>,
```
[the remark this answers is not in the document]
Comment (rina23q) [reply crates/core/c8y_api/src/json_c8y.rs:258]: Good point. It can be a flag `clear: true` for example. I'll consider.

[the remark this answers is not in the document]
Comment (didier-wenzek) [reply crates/core/c8y_api/src/json_c8y.rs:258]: @albinsuresh at first, I was aligned with you, asking @rina23q to group all related data into a single `struct`.

But I have to acknowledge that there is a nice property of this two-parts message. This gives a nice way to represent either an alarm or a message to clear the alarm. There is a similar issue with commands, where one needs to be cautious before [de-serializing a command message](https://github.com/thin-edge/thin-edge.io/blob/main/crates/extensions/c8y_mapper_ext/src/converter.rs#L740) because the command structs are enable to represent the clearing case.

File: crates/core/tedge_api/src/alarm.rs (hunk 6 above)
Comment (rina23q) [reply crates/core/tedge_api/src/alarm.rs:14]: I'm not sure if I understood @albinsuresh here correctly. So what I wanted to say here are
1. `Option<String>` is wrong. It should be `String` as @didier-wenzek suggested.
2. The value of `alarm_type` is always delivered by [Channel](https://github.com/thin-edge/thin-edge.io/blob/c0ded23e1927a156f3f83941eda8fc854c6c997f/crates/core/tedge_api/src/mqtt_topics.rs#L323). Since our MQTT scheme allows an empty string as an alarm type (at least this `Channel` implementation returns `String("")`), we need to remember this `alarm_type` may have `String("")`. 
3. If we don't want to allow `String("")`, the implementation should be done in the `Channel` level.

File: crates/extensions/c8y_mapper_ext/src/alarm_converter.rs (hunk 8 above)
Comment (rina23q) [reply crates/extensions/c8y_mapper_ext/src/alarm_converter.rs:65]: Just not that `impl TryFrom<&Message> for ThinEdgeAlarm` is not ideal as `alarm_type` and `source` are given by the upper level. Re-invention of Message to `alarm_type` and `source` are not smart.

Here we need 3 things to create `ThinEdgeAlarm`.
1. `alarm_type` (originally from topic, but here derived from upper level)
2. `source` (originally from topic, but here derived from upper level)
3. `mqtt_payload` (get from `Message`).

Technically we can create a generic struct and implement `TryFrom` for the struct as these 3 parameters are the same for measurements and events. However, I'd address this level of sophistication later.

In short, I will create a function which takes the 3 fields and return `Result<ThinEdgeAlarm, Err>` in this PR.

File: crates/extensions/c8y_mapper_ext/src/alarm_converter.rs (hunk 5 above)
Comment (rina23q) [reply crates/extensions/c8y_mapper_ext/src/alarm_converter.rs]: Done with ec55b7328b99294efd1e2e16b7430bb7bed5a371

File: crates/core/tedge_api/src/alarm.rs (hunk 6 above)
Comment (rina23q) [reply crates/core/tedge_api/src/alarm.rs]: Done with ec55b7328b99294efd1e2e16b7430bb7bed5a371

File: crates/core/tedge_api/src/alarm.rs [hunk 14]
```diff
@@ -64,316 +90,117 @@ pub enum ThinEdgeJsonDeserializerError {
...
-            AlarmSeverity::Critical => write!(f, "CRITICAL"),
-            AlarmSeverity::Major => write!(f, "MAJOR"),
-            AlarmSeverity::Minor => write!(f, "MINOR"),
-            AlarmSeverity::Warning => write!(f, "WARNING"),
-        }
-    }
-}
-
-impl ThinEdgeAlarm {
-    pub fn try_from(
```
[the remark this answers is not in the document]
Comment (rina23q) [reply crates/core/tedge_api/src/alarm.rs:95]: Done with ec55b7328b99294efd1e2e16b7430bb7bed5a371

File: crates/extensions/c8y_mapper_ext/src/alarm_converter.rs (hunk 8 above)
Comment (rina23q) [reply crates/extensions/c8y_mapper_ext/src/alarm_converter.rs]: Done with ec55b7328b99294efd1e2e16b7430bb7bed5a371

File: crates/core/c8y_api/src/json_c8y.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (rina23q) [reply crates/core/c8y_api/src/json_c8y.rs:258]: I think @albinsuresh and @didier-wenzek are on the same page here. But I believe that @albinsuresh says `C8yCreateAlarm` doesn't need to have a sub-struct as it's used only for **serializing** a struct to c8y's way.

In c8y's API, [`"status": "CLEARED"`](https://cumulocity.com/guides/10.4.6/reference/alarms/#alarm-applicationvndcomnsncumulocityalarmjson) is the key. We don't need to use it now as we clear an alarm by SmartREST, however, once we move to JSON over MQTT, we will need this field. 

So I'll change the current struct to align more with c8y.

File: tests/RobotFramework/tests/cumulocity/telemetry/thin-edge_device_telemetry.robot [hunk 15]
```diff
@@ -69,12 +69,50 @@ Thin-edge devices support sending custom events without type in topic
...
+    Log    ${alarms}
+
+
+Thin-edge devices support sending custom alarms without severity in payload
+    Execute Command    tedge mqtt pub te/device/main///a/myCustomAlarmType2 '{ "text": "Some test alarm" }'
+    ${alarms}=    Device Should Have Alarm/s    expected_text=Some test alarm    severity=MINOR    minimum=1    maximum=1    type=myCustomAlarmType2
+    Log    ${alarms}
+
+
+Thin-edge devices support sending custom alarms with unknown severity in payload
```
[the remark this answers is not in the document]
Comment (reubenmiller) [reply tests/RobotFramework/tests/cumulocity/telemetry/thin-edge_device_telemetry.robot:96]: To be honest, I'm ok with a more lenient way handling the severity (e.g. using a default value when an invalid value is given).

File: crates/core/c8y_api/src/json_c8y.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (albinsuresh) [reply crates/core/c8y_api/src/json_c8y.rs:350]: Aah okay. As long as there's an outer `From` that the consumer can directly use, that's sufficient. I missed that there was one, wrapping these.

File: crates/core/c8y_api/src/json_c8y.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (albinsuresh) [reply crates/core/c8y_api/src/json_c8y.rs:258]: @didier-wenzek I was also fully aligned with you on that decomposition of the `ThinEdgeAlarm` struct, precisely for that reason that you highlighted. That's why even the former struct had this decomposition using a sub-struct representing the payload. But, here, as @rina23q also highlighted, my concern was about replicating the same style for the `C8yCreateAlarm` struct, which looked unnecessary, as the representation of a "clear alarm message" was not an empty sub-struct in C8y.

IMO, instead of overloading the `C8yCreateAlarm` struct, by making `severity` and `text` fields optional, to handle clearing as well, it'd be better have an entirely different struct, a `C8yClearAlarm`, to represent that case.

File: tests/RobotFramework/tests/cumulocity/telemetry/thin-edge_device_telemetry.robot (hunk 11 above)
[the remark this answers is not in the document]
Comment (albinsuresh) [reply tests/RobotFramework/tests/cumulocity/telemetry/thin-edge_device_telemetry.robot:78]: Aah okay. I thought that rule applied only for measurements as the old regime did not support types in the topic, where-as it has always been mandatory for alarms and events. But, I guess this is fine for the sake of consistency.

[the remark this answers is not in the document]
Comment (rina23q) [reply tests/RobotFramework/tests/cumulocity/telemetry/thin-edge_device_telemetry.robot:78]: @albinsuresh Yes, I would really recommend that we should keep the consistency in all measurements, events, and alarms at this PR (the main goal is to support the new "te" topic), then later change/unchange all of them together if we want to change the detail level decisions..

File: crates/core/c8y_api/src/json_c8y.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (rina23q) [reply crates/core/c8y_api/src/json_c8y.rs:258]: > as the representation of a "clear alarm message" was not an empty sub-struct in C8y.

Correct. I start thinking rather making two different struct for creating an alarm and clearing an alarm is the right way. Why?
1. As @albinsuresh said, "clearing an alarm" is not an empty sub-struct. It has ["status": "CLEARED"](https://cumulocity.com/guides/10.4.6/reference/alarms/#alarm-applicationvndcomnsncumulocityalarmjson) or not.
2. But as I said, having **optional** `severity`, `text` and `time` are also wrong as they are mandatory to create an alarm.  

So I start feeling we are talking about two different (but similar in terms of alarm) data models and trying to keep it in one struct. I would go to have two structs if others agree with me. Otherwise, I will not change.

File: crates/core/c8y_api/src/json_c8y.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (rina23q) [reply crates/core/c8y_api/src/json_c8y.rs:350]: Exactly. As the test shows ;)

```rust
let c8y_alarm = C8yCreateAlarm::from(&tedge_alarm);
```

File: crates/core/c8y_api/src/smartrest/alarm.rs [hunk 16]
```diff
@@ -1,168 +1,136 @@
...
+use crate::json_c8y::C8yCreateAlarm;
 use time::format_description::well_known::Rfc3339;
-use time::OffsetDateTime;
 
-/// Converts from thin-edge alarm to C8Y alarm SmartREST message
-pub fn serialize_alarm(alarm: ThinEdgeAlarm) -> Result<String, time::error::Format> {
-    match alarm.data {
-        None => Ok(format!("306,{}", alarm.name)),
+/// Serialize C8yCreateAlarm to SmartREST message
+pub fn serialize_alarm(alarm: C8yCreateAlarm) -> Result<String, time::error::Format> {
```
[the remark this answers is not in the document]
Comment (albinsuresh) [reply crates/core/c8y_api/src/smartrest/alarm.rs:6]: Creating the SmartREST alarm directly from the `ThinEdgeAlarm` was intentional in the old regime, to avoid the creation of intermediate structs. That's why necessary defaults were also applied at this level itself, when the input didn't have some. But, now that you've made this change and the plan is to migrate JSON over MQTT anyway, which would need this `C8yCreateAlarm` struct anyway, let it be.

File: crates/core/c8y_api/src/json_c8y.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (rina23q) [reply crates/core/c8y_api/src/json_c8y.rs:258]: Discussed with @albinsuresh, and this model satisfies both 1) and 2) as I mentioned in the above comment. I'll change.

```rust
enum C8yAlarm {
    Create(C8yCreateAlarm),
    Clear(C8yClearAlarm)
    
}

struct C8yCreateAlarm {
    // all necessary fields, and all are mandatory.
    
}

struct C8yClearAlarm {
    alarm_type: &str,
    source: Option<SourceInfo>,
}
```

```rust
let c8y_alarm = C8yAlarm::from(&tedge_alarm);
```

File: crates/extensions/c8y_mapper_ext/src/converter.rs (hunk 9 above)
Comment (rina23q) [reply crates/extensions/c8y_mapper_ext/src/converter.rs]: As @didier-wenzek, changing from `EntityMetadata` to `EntityTopicId` solved this. https://github.com/thin-edge/thin-edge.io/pull/2267/commits/6536b54e6401841d88d0b20c3e24650ff8d9e894

File: crates/extensions/c8y_mapper_ext/src/error.rs [hunk 17]
```diff
@@ -115,6 +115,9 @@ pub enum ConversionError {
 
     #[error(transparent)]
     FileError(#[from] FileError),
+
+    #[error(transparent)]
+    FileC8yAlarmError(#[from] c8y_api::json_c8y::C8yAlarmError),
```
Comment (didier-wenzek) [inline crates/extensions/c8y_mapper_ext/src/error.rs:120]: Why a `File` prefix?

File: crates/core/c8y_api/src/smartrest/topic.rs [hunk 18]
```diff
@@ -101,6 +102,21 @@ impl From<&EntityMetadata> for C8yTopic {
...
+                None => Self::SmartRestResponse,
+                Some(info) => Self::ChildSmartRestResponse(info.id.clone()),
+            },
+            C8yAlarm::Clear(alarm) => match alarm.source.as_ref() {
+                None => Self::SmartRestResponse,
+                Some(info) => Self::ChildSmartRestResponse(info.id.clone()),
+            },
+        }
+    }
+}
```
Comment (didier-wenzek) [inline crates/core/c8y_api/src/smartrest/topic.rs:118]: This will have to be updated, once merged this PR: https://github.com/thin-edge/thin-edge.io/pull/2266#discussion_r1330224805

File: crates/core/c8y_api/src/json_c8y.rs [hunk 19]
```diff
@@ -286,41 +288,47 @@ pub struct C8yClearAlarm {
...
+                    source,
+                    severity: C8yCreateAlarm::convert_severity(tedge_alarm_data),
+                    text: C8yCreateAlarm::convert_text(tedge_alarm_data, &alarm_type),
+                    time: C8yCreateAlarm::convert_time(tedge_alarm_data),
+                    fragments: C8yCreateAlarm::convert_extras(tedge_alarm_data),
+                }),
+            };
+            Ok(c8y_alarm)
+        } else {
+            Err(C8yAlarmError::UnsupportedExternalDeviceId(
```
Comment (didier-wenzek) [inline crates/core/c8y_api/src/json_c8y.rs:313]: The error type is not appropriate: the source is device topic identifier not an external device id.

File: crates/core/c8y_api/src/json_c8y.rs [hunk 20]
```diff
@@ -236,96 +240,184 @@ impl SourceInfo {
...
-            Some(data) => {
-                text = data.text.clone().unwrap_or_else(|| alarm_type.clone());
-                time = data.time.unwrap_or_else(OffsetDateTime::now_utc);
-                fragments = data.alarm_data.clone();
-            }
+    fn convert_source(entity: &EntityMetadata) -> Option<SourceInfo> {
+        match entity.r#type {
+            EntityType::MainDevice => None,
+            EntityType::ChildDevice => Some(make_c8y_source_fragment(&entity.entity_id.clone())),
+            EntityType::Service => None, // TODO!: What is for service?
```
Comment (didier-wenzek) [inline crates/core/c8y_api/src/json_c8y.rs:323]: A service is addressed the same way as for a child-device.

File: crates/extensions/c8y_mapper_ext/src/converter.rs [hunk 21]
```diff
@@ -318,34 +318,20 @@ impl CumulocityConverter {
...
-            }
-            mqtt_messages.append(&mut messages);
-            Ok(mqtt_messages)
-        } else if topic.name.starts_with(INTERNAL_ALARMS_TOPIC) {
-            self.alarm_converter.process_internal_alarm(message);
-            Ok(vec![])
-        } else {
-            Err(ConversionError::UnsupportedTopic(topic.name.clone()))
-        }
+        self.size_threshold.validate(input)?;
```
Comment (didier-wenzek) [inline crates/extensions/c8y_mapper_ext/src/converter.rs:325]: This check can be removed. Only the size of output messages must be checked.

**Review by didier-wenzek — COMMENTED** [review]
Really close to be merged. However, I still have some questions and comments.

File: crates/extensions/c8y_mapper_ext/src/error.rs (hunk 17 above)
Comment (rina23q) [reply crates/extensions/c8y_mapper_ext/src/error.rs:120]: Lol. I thought it's `From` :D:D

File: crates/core/c8y_api/src/smartrest/topic.rs (hunk 18 above)
Comment (rina23q) [reply crates/core/c8y_api/src/smartrest/topic.rs:118]: Also, alarm won't use this after #2242 is done.

File: crates/core/c8y_api/src/json_c8y.rs (hunk 20 above)
Comment (rina23q) [reply crates/core/c8y_api/src/json_c8y.rs]: Done with 816a8cb06525fbaef157ef25690ba983ed647949

File: crates/core/c8y_api/src/json_c8y.rs (hunk 19 above)
Comment (rina23q) [reply crates/core/c8y_api/src/json_c8y.rs]: Done with 816a8cb06525fbaef157ef25690ba983ed647949

File: crates/extensions/c8y_mapper_ext/src/error.rs (hunk 17 above)
Comment (rina23q) [reply crates/extensions/c8y_mapper_ext/src/error.rs]: Fix with 816a8cb06525fbaef157ef25690ba983ed647949

File: crates/extensions/c8y_mapper_ext/src/converter.rs (hunk 21 above)
Comment (rina23q) [reply crates/extensions/c8y_mapper_ext/src/converter.rs:325]: I tried to move this logic after creating either SmartREST or JSON over MQTT message, but one unit test failed. Therefore, I'll leave it as it is now. 

P.S. If the input JSON is way too big, most likely the output mqtt message is also too big. So, checking input is not completely useless.

Comment (didier-wenzek) [reply crates/extensions/c8y_mapper_ext/src/converter.rs:325]: Okay, we will see later.

Comment (albinsuresh) [reply crates/extensions/c8y_mapper_ext/src/converter.rs:325]: > If the input JSON is way too big, most likely the output mqtt message is also too big. So, checking input is not completely useless.

Yes, that was the reason why it was done, to avoid an unnecessary conversion of a huge input that'd be thrown away a few steps later.

**Review by albinsuresh — APPROVED** [review]
